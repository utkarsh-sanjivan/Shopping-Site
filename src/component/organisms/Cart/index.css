.cart-container {
    width: 30%;
    border: 1px solid #dedede;
    padding: 20px 10px;
    margin: 19px 1%;
}

@media only screen and (max-width: 700px) {
    .cart-container {
        width: 50%;
    }
}

@media only screen and (max-width: 400px) {
    .cart-container {
        width: 80%;
        margin: 10px auto;
    }
}

.cart-header {
    display: flex;
    line-height: 40px;
    justify-content: space-between;
}

.header-item {
    width: 25%;
    text-align: center;
}

.cart-footer {
    
}

.footer-item {
    text-align: right;
}