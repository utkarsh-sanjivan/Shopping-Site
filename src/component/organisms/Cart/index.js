import React from 'react';
import './index.css';
import CartItem from '../../molecules/CartItem';

export default function Cart(props) {
    return (
        <div className='cart-container'>
            <div className='cart-header'>
                <div className='header-item'>Product</div>
                <div className='header-item'>Price</div>
                <div className='header-item'>Quantity</div>
                <div className='header-item'>Total</div>
            </div>
            {props.cartItem.map((item, index) => 
                <CartItem 
                    {...item}
                    index={index}
                    key={props.id}
                    removeOne={props.removeOne}
                    addOne={props.addOne}
                />
            )}
            {props.cartItem.length>0?
                <div className='cart-footer'>
                    <hr />
                    <div className='footer-item'>Total: â‚¹{props.totalAmount}</div>
                </div>
            : null}
        </div>
    )
} 