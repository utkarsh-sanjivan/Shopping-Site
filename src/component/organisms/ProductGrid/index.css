.product-grid {
    width: 70%;
    display: grid;
    grid-template-columns: auto auto auto auto;
    padding: 10px;
    gap: 10px;
}

@media only screen and (max-width: 1400px) {
    .product-grid {
        grid-template-columns: auto auto auto;
    }
}

@media only screen and (max-width: 900px) {
    .product-grid {
        grid-template-columns: auto auto;
    }
}

@media only screen and (max-width: 700px) {
    .product-grid {
        width: 50%;
        grid-template-columns: auto;
    }
}

@media only screen and (max-width: 400px) {
    .product-grid {
        margin: auto;
        width: 80%;
    }
}