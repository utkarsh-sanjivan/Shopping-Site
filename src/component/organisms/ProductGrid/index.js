import React from 'react';
import GridItem from './../../molecules/GridItem/index';
import './index.css';

export default function ProductGrid(props) {
    /*
        Grid Component that holds a particular product.
        It has all the details of product along with price and quantity.
        It also has a 'Add To Cart' button that can be used to add the product to the cart.
    */
    return (
        <div className='product-grid'>
            {props.productList.map(product => 
                <GridItem
                    {...product}
                    key={product.id}
                    handleSelect={props.handleSelect}
                />
            )}
        </div>
    )
}