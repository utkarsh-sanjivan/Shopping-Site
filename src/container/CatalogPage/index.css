.body {
    display: flex;
    align-items: flex-start;
    min-height: 72vh;
}

@media only screen and (max-width: 500px) {
    .body {
        flex-direction: column-reverse;
    }
}