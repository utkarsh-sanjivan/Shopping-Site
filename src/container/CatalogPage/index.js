import React, { useState, useEffect } from 'react';
import ProductGrid from './../../component/organisms/ProductGrid/index';
import Cart from './../../component/organisms/Cart/index';
import './index.css';
import axios from './../../utils/axios';

export default function CatalogPage(props) {
    const { setTotalProductCount, totalProductCount } = props;
    const [productList, setProductList] = useState([]);
    const [cartItem, setCartItem] = useState([]);
    const [totalAmount, setTotalAmount] = useState(0);

    useEffect(() => {
        axios.get('aea5d98a-654d-4423-bd99-6fbb90843730')
            .then(data => {
                setProductList(data.data.data);
            })
    }, []);

    /* 
        Function to select a product from the product grid.
        Takes in the the id for the product.
        If the product doesnot exits then add the product to cart and reduce the count in the product grid.
        If the product exists in the cart then increase the count in cart and decrease in the product grid.
    */
    const handleSelect = currentProductId => {
        let cartIndex = cartItem.findIndex(val => val.id===currentProductId);
        if (cartIndex>-1) {
            let tempCartItem = [ ...cartItem ];
            setTotalProductCount(tempCartItem[cartIndex].quantity===0? totalProductCount :totalProductCount+1);
            setTotalAmount(tempCartItem[cartIndex].quantity===0? totalAmount: totalAmount+parseInt(tempCartItem[cartIndex].price));
            tempCartItem[cartIndex].cartQuantity+=tempCartItem[cartIndex].quantity===0? 0: 1;
            tempCartItem[cartIndex].quantity-=tempCartItem[cartIndex].quantity===0? 0 :1;
            setCartItem(tempCartItem);
        } else {
            let tempProduct = [];
            let tempCart = {};
            productList.forEach(val => {
                if(val.id===currentProductId) {
                    val = { ...val, quantity: val.quantity-1, cartQuantity: 1 };
                    tempCart=val;
                    setTotalAmount(totalAmount+parseInt(val.price));
                }
                tempProduct.push(val);
            });
            setTotalProductCount(totalProductCount+1);
            setProductList(tempProduct);
            setCartItem([
                ...cartItem,
                tempCart,
            ]);
        }
    }

    /*
        Function to add product in the cart.
        Takes in the the id for the product.        
    */
    const addOne = currentProductId => {
        let tempProduct = [ ...productList ];
        tempProduct = tempProduct.map(product => {
            if (product.id === currentProductId) {
                return { 
                    ...product,
                    quantity: product.quantity===0? product.quantity :product.quantity-1,
                    cartQuantity: product.quantity===0? product.cartQuantity :product.cartQuantity+1
                };
            }
            return product;
        })
        let tempCartItem = cartItem.map(val => {
            if(val.id===currentProductId) {
                if (val.quantity>0) setTotalAmount(totalAmount+parseInt(val.price));
                setTotalProductCount(val.quantity===0? totalProductCount :totalProductCount+1);
                return val.quantity===0?
                    val
                    : { ...val, quantity: val.quantity-1, cartQuantity: val.cartQuantity+1 }
            }
            return val;
        });
        setProductList(tempProduct);
        setCartItem(tempCartItem);
    }

    /*
        Function to remove product in the cart.
        Takes in the the id for the product.
        If the product count in cart is 1 then remove the product from the cart.
        If the product count in the cart is more than 1 then just decrease the count by 1..
    */
    const removeOne = (currentProductId, currentProductQuantity, currentProductIndex) => {
        let tempCartItem = [ ...cartItem ];
        if (parseInt(currentProductQuantity) === 1) {
            setProductList(productList.map(product => {
                if (product.id===currentProductId) {
                    return { ...product, quantity: product.quantity+1, cartQuantity: product.cartQuantity-1 }
                }
                return product;
            }));
            setTotalAmount(totalAmount-parseInt(tempCartItem[currentProductIndex].price));
            tempCartItem.splice(currentProductIndex, 1);
            setCartItem(tempCartItem);
        } else {
            let tempProduct = [ ...productList ];
            tempProduct = tempProduct.map(product => {
                if (product.id === currentProductId) {
                    return { 
                        ...product,
                        quantity: product.cartQuantity===0? product.quantity :product.quantity+1,
                        cartQuantity: product.cartQuantity===0? product.cartQuantity :product.cartQuantity-1
                    };
                }
                return product;
            })
            let tempCartItem = cartItem.map(val => {
                if(val.id===currentProductId) {
                    if (val.cartQuantity>0) setTotalAmount(totalAmount-parseInt(val.price));
                    return val.cartQuantity===0?
                        val
                        : { ...val, quantity: val.quantity+1, cartQuantity: val.cartQuantity-1 }
                }
                return val;
            });
            setProductList(tempProduct);
            setCartItem(tempCartItem);
        }
        setTotalProductCount(totalProductCount-1);
    }

    /*
        Main body of the shopping catalog which contain two parts.
        Product Grid -> Contains all the product in the for of grid that has the detail of the product as well as the 'Add to Cart' button.
        Cart -> Cart component that contains all the selected products, their price and their count.
    */
    return (
        <div className='body'>
            <ProductGrid 
                productList={productList}
                handleSelect={handleSelect}
            />
            <Cart 
                cartItem={cartItem}
                addOne={addOne}
                removeOne={removeOne}
                totalAmount={totalAmount}
            />
        </div>
    );
}