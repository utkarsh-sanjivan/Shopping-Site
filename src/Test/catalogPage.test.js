import React from 'react';
import { create } from "react-test-renderer";
import CatalogPage from './../container/CatalogPage/index';
import ProductGrid from './../component/organisms/ProductGrid/index';
import Cart from './../component/organisms/Cart/index';

let wrapper;
let root;
const renderComponent = props => <CatalogPage {...props} />;
const baseProps = {
    totalProductCount: 0,
    setTotalProductCount: () => {},
};

beforeEach(() => {
  wrapper = create(renderComponent(baseProps));
  root=wrapper.root;
});

describe('Testing Catalog Page with 0 product count', () => {
    it('Catalog Page should render properly', () => {
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    // Test for numbers of div DOM element when no products are available in the cart.
    it('should atleast 8 div component', () => {
        expect(root.findAllByType('div').length).toEqual(8);
    });

    // Test for numbers of Cart Component when no products are available in the cart.
    it('should atleast Cart component', () => {
        expect(root.findAllByType(Cart).length).toEqual(1);
    });

    // Test for numbers of Product Grid Component when no products are available in the cart.
    it('should atleast 1 Product Grid component', () => {
        expect(root.findAllByType(ProductGrid).length).toEqual(1);
    });
});