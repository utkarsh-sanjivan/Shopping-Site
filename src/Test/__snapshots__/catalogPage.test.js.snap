// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Catalog Page with 0 product count Catalog Page should render properly 1`] = `
<div
  className="body"
>
  <div
    className="product-grid"
  />
  <div
    className="cart-container"
  >
    <div
      className="cart-header"
    >
      <div
        className="header-item"
      >
        Product
      </div>
      <div
        className="header-item"
      >
        Price
      </div>
      <div
        className="header-item"
      >
        Quantity
      </div>
      <div
        className="header-item"
      >
        Total
      </div>
    </div>
  </div>
</div>
`;
